### DISCUSSION TITLE
How to pass state to the suffix command

### DISCUSSION THREAD

**Discussion by cperivol** [magit/transient#298] [Support] [answered]
Hi all!

I am trying to pass some information from the prefix to the suffix command, ideally by binding the suffix command arguments. For example I would like to do something like this:

```elisp
(defun action-command (buffer)
  (interactive "bBuffer: ")
  (do-something buffer))

(transient-define-prefix popup-foo ()
  "This calls actions and passes some stuff to them"
  ["Actions" ("a" "Some action" action-command)]
  (transient-setup 'popup-foo :arg (find-a-buffer-at-popup-time)))
  ```
  
  I am suspecting that I am misunderstanding something about how transient is supposed to be used, can someone help me clarify the situation?
  Thanks!

**Comment by cperivol** [comment]
My current "solution" is this abomination:

```elisp
(defun cperivol--transient-init-objects (&rest args)
  (when (and (boundp 'cperivol-transient--caller) (functionp cperivol-transient--caller))
    (dolist (obj transient--suffixes)
      (setf (oref obj command)
            (let ((fn (gensym)))
              (eval
               `(defun ,fn ()
                  (interactive)
                  (funcall ,cperivol-transient--caller (quote ,(oref obj command)))))
              fn)))))

(advice-add #'transient--init-objects :after #'cperivol--transient-init-objects)

(transient-define-prefix foo-popup ()
  "Doc."
  ["Actions"
   ("a" "Action" action-command)]
  (interactive "P")
  (let* ((buf (get-the-buffer-at-popup-time))
	 (dir (get-the-dir-at-popup-time))
         (cperivol-transient--caller (apply-partially
                                      (lambda (dir buf fn)
                                        (let ((default-directory dir))
                                          (funcall fn buf)))
                                      dir buf)))
    (transient-setup 'foo-popup)))
```

**Comment by tarsius** [comment, marked as answer]
I'm not sure what you are trying to do but I am guessing its similar to what `magit-configure-branch` does, just for a buffer instead of a branch. Look at that and some of its suffixes for inspiration.

**Reply by cperivol** [reply]
Thanks! That seems to be it. I will try it tomorrow but for posterity the trick seems to be that you need to wrap the suffix command in  `transient-define-suffix` and if you do that you can have access to the `transient-current-prefix` which is otherwise `nil`. Now that you can get the prefix you can get the scope!

**Reply by tarsius** [reply]
`transient-current-prefix` is non-nil whenever a command is invoked from a prefix menu. It does not necessarily have to be defined using `transient-define-suffix` (but it may make sense to do so anyway).

**Reply by cperivol** [reply]
Nice! It worked. For some reason if you edebug the suffix command though, then `transient-current-prefix` winds up being nil, that's why I was confused before.
